### PR TITLE
Extraction worldtube communication actions

### PR DESCRIPTION
**Pull request by moxcodes**
## Proposed changes

Add in the actions necessary to get the data from the worldtube component to the evolution component in CCE.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Dependencies
- [x] #1931

### CODE REVIEW

**Comment by moxcodes** [comment]
rebased on develop, no longer dependent.

File: src/Evolution/Systems/Cce/Actions/InitializeCharacteristicEvolution.hpp [hunk 1]
```diff
@@ -217,45 +215,46 @@ struct InitializeCharacteristicEvolution {
...
                     const ParallelComponent* const /*meta*/) noexcept {
+    const double initial_time_value =
+        db::get<InitializationTags::StartTime>(box);
+    const double step_size = db::get<InitializationTags::TargetStepSize>(box);
+
     auto initialize_box = Initialization::merge_into_databox<
         InitializeCharacteristicEvolution,
         db::AddSimpleTags<Tags::BoundaryTime>, db::AddComputeTags<>>(
-        std::move(box), std::numeric_limits<double>::quiet_NaN());
+        std::move(box), TimeStepId{true, -1, Time{Slab{-1.0, 0.0}, {0, 1}}});
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/InitializeCharacteristicEvolution.hpp]: why are you putting in a fake TimeStepId?

File: tests/Unit/Evolution/Systems/Cce/Actions/Test_InitializeCharacteristicEvolution.cpp [hunk 2]
```diff
@@ -140,6 +140,9 @@ SPECTRE_TEST_CASE(
   const double frequency = 0.1 * value_dist(gen);
   const double amplitude = 0.1 * value_dist(gen);
   const double target_time = 50.0 * value_dist(gen);
+  if (file_system::check_if_file_exists(filename)) {
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/Cce/Actions/Test_InitializeCharacteristicEvolution.cpp:143]: can you factor this into a separate commit

File: src/Evolution/Systems/Cce/Actions/RequestBoundaryData.hpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        Actions::BoundaryComputeAndSendToEvolution<EvolutionComponent>>(
+        Parallel::get_parallel_component<WorldtubeBoundaryComponent>(cache),
+        db::get<::Tags::TimeStepId>(box));
+    return std::forward_as_tuple(std::move(box));
+  }
+};
+
+/*!
+ * \ingroup ActionsGroup
+ * \brief Requests boundary data be sent from `WorldtubeBoundaryComponent` to
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/RequestBoundaryData.hpp:66]: give the full namespace so they will make links

File: src/Evolution/Systems/Cce/Actions/RequestBoundaryData.hpp [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+ * `EvolutionComponent`.
+ *
+ * \details Calls the simple action
+ * `Cce::Actions::BoundaryComputeAndSendToEvolution<EvolutionComponent>` on the
+ * `WorldtubeBoundaryComponent`, which performs boundary computations then sends
+ * data to the `EvolutionComponent` via `Cce::Actions::ReceiveWorldtubeData`.
+ * Requests the current `Tags::TimeStepId`.
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/RequestBoundaryData.hpp]: Request data at the current ?

File: src/Evolution/Systems/Cce/Actions/RequestBoundaryData.hpp [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+/*!
+ * \ingroup ActionsGroup
+ * \brief Requests boundary data be sent from `WorldtubeBoundaryComponent` to
+ * `EvolutionComponent`.
+ *
+ * \details Calls the simple action
+ * `Cce::Actions::BoundaryComputeAndSendToEvolution<EvolutionComponent>` on the
+ * `WorldtubeBoundaryComponent`, which performs boundary computations then sends
+ * data to the `EvolutionComponent` via `Cce::Actions::ReceiveWorldtubeData`.
+ * Requests the `Tags::Next<Tags::TimeStepId>` (for the next timestep).
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/RequestBoundaryData.hpp]: Request data at the ?

File: src/Evolution/Systems/Cce/Actions/RequestBoundaryData.hpp [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+namespace Cce {
+namespace Actions {
+/// \cond
+template <typename EvolutionComoponent>
+struct BoundaryComputeAndSendToEvolution;
+/// \endcond
+
+/*!
+ * \ingroup ActionsGroup
+ * \brief Requests boundary data be sent from `WorldtubeBoundaryComponent` to
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/RequestBoundaryData.hpp:23]: namespace so links will show up in the documentation

File: tests/Unit/Evolution/Systems/Cce/Actions/Test_RequestBoundaryData.cpp [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+template <typename EvolutionComponent>
+struct MockBoundaryComputeAndSendToEvolution {
+  template <typename ParallelComponent, typename... DbTags,
+            typename Metavariables, typename ArrayIndex,
+            Requires<tmpl2::flat_any_v<cpp17::is_same_v<
+                ::Tags::Variables<
+                    typename Metavariables::cce_boundary_communication_tags>,
+                DbTags>...>> = nullptr>
+  static void apply(db::DataBox<tmpl::list<DbTags...>>& /*box*/,
+                    Parallel::ConstGlobalCache<Metavariables>& /*cache*/,
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/Cce/Actions/Test_RequestBoundaryData.cpp]: const the unused variables

File: tests/Unit/Evolution/Systems/Cce/Actions/Test_RequestBoundaryData.cpp [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+
+  MAKE_GENERATOR(gen);
+  UniformCustomDistribution<double> value_dist{0.1, 0.5};
+  // first prepare the input for the modal version
+  const double mass = value_dist(gen);
+  const std::array<double, 3> spin{
+      {value_dist(gen), value_dist(gen), value_dist(gen)}};
+  const std::array<double, 3> center{
+      {value_dist(gen), value_dist(gen), value_dist(gen)}};
+  gr::Solutions::KerrSchild solution{mass, spin, center};
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/Cce/Actions/Test_RequestBoundaryData.cpp]: const

File: tests/Unit/Evolution/Systems/Cce/Actions/Test_RequestBoundaryData.cpp [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+  runner.set_phase(test_metavariables::Phase::Initialization);
+  ActionTesting::emplace_component<evolution_component>(
+      &runner, 0, start_time,
+      InitializationTags::EndTime::create_from_options(end_time, filename),
+      target_step_size);
+  ActionTesting::emplace_component<worldtube_component>(
+      &runner, 0,
+      InitializationTags::H5WorldtubeBoundaryDataManager::create_from_options(
+          l_max, filename, buffer_size,
+          std::make_unique<intrp::BarycentricRationalSpanInterpolator>(3u,
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/Cce/Actions/Test_RequestBoundaryData.cpp]: 3_st, 4_st

File: src/Evolution/Systems/Cce/Actions/ReceiveWorldtubeData.hpp [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+ * \brief Takes the boundary data needed to perform the CCE linear solves as
+ * arguments and puts them in the \ref DataBoxGroup, updating the
+ * `Cce::Tags::BoundaryTime` accordingly.
+ *
+ * \details The boundary data is computed by a separate component, and packaged
+ * into a `Variables<tmpl::list<BoundaryTags...>>` which is sent in the argument
+ * of the simple action invocation. The `TimeStepId` is also provided to confirm
+ * the time associated with the passed boundary data.
+ *
+ * \note this action re-activates the algorithm by `perform_algorithm()` in case
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/ReceiveWorldtubeData.hpp]: don't call `perform_algorithm` within an action; Wanting to do so means that this action should not be a simple_action.  Just promote this to a iterable action by taking the `is_ready` check from `BlockUntilBoundaryDataReceived` and get rid of `BlockUntilBoundaryDataReceived`

File: src/Evolution/Systems/Cce/Actions/BoundaryComputeAndSendToEvolution.hpp [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+                    typename Metavariables::cce_boundary_communication_tags>,
+                DbTags>...>> = nullptr>
+  static void apply(db::DataBox<tmpl::list<DbTags...>>& box,
+                    Parallel::ConstGlobalCache<Metavariables>& cache,
+                    const ArrayIndex& /*array_index*/,
+                    const TimeStepId& time) noexcept {
+    if (not db::get<Tags::H5WorldtubeBoundaryDataManager>(box)
+                .populate_hypersurface_boundary_data(
+                    make_not_null(&box), time.substep_time().value())) {
+      Parallel::printf(
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/BoundaryComputeAndSendToEvolution.hpp]: Just use `ERROR`

File: src/Evolution/Systems/Cce/Actions/BoundaryComputeAndSendToEvolution.hpp [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+  }
+
+ private:
+  template <typename ProxyType, typename VariableTagList, typename... Tags>
+  static void send_boundary_data(
+      ProxyType& extractor_proxy,
+      const Variables<VariableTagList>& boundary_quantities,
+      const TimeStepId& time, tmpl::list<Tags...> /*meta*/) noexcept {
+    Parallel::simple_action<
+        Actions::ReceiveWorldtubeData<EvolutionComponent, tmpl::list<Tags...>>>(
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/BoundaryComputeAndSendToEvolution.hpp]: just send the variables and get rid of this helper function.  you can unpack on the receiving side if necessary

**Review by kidder — CHANGES_REQUESTED** [review]

File: src/Evolution/Systems/Cce/Actions/ReceiveWorldtubeData.hpp (hunk 10 above)
Comment (moxcodes) [reply src/Evolution/Systems/Cce/Actions/ReceiveWorldtubeData.hpp]: Sorry, I need more information to work out what to do on this comment, because this simple action performs the data transfer via the arguments. Are you suggesting that I use the `receive_data` and inbox mechanisms instead of using the simple action, and examine the inboxes to determine is_ready?

Comment (kidder) [reply src/Evolution/Systems/Cce/Actions/ReceiveWorldtubeData.hpp]: yes, that is what we are saying...

**Comment by kidder** [comment]
reviewed with @nilsdeppe

File: src/Evolution/Systems/Cce/Actions/RequestBoundaryData.hpp (hunk 6 above)
Comment (moxcodes) [reply src/Evolution/Systems/Cce/Actions/RequestBoundaryData.hpp:23]: These refer to the template parameters of this struct.

**Comment by moxcodes** [comment]
fixups posted.
new non-fixup commits are `remove Cce test file before test` and `Add receive_data in action testing proxy 'array'`
Cheers!

File: src/Evolution/Systems/Cce/Actions/ReceiveWorldtubeData.hpp [hunk 13]
```diff
@@ -27,52 +29,97 @@ namespace Actions {
...
+                    const ParallelComponent* const /*meta*/) noexcept {
+    auto& inbox = tuples::get<Cce::ReceiveTags::BoundaryData<
+        typename Metavariables::cce_boundary_communication_tags>>(inboxes);
+    tmpl::for_each<typename Metavariables::cce_boundary_communication_tags>(
+        [&inbox, &box ](auto tag_v) noexcept {
+          using tag = typename decltype(tag_v)::type;
+          db::mutate<tag>(
+              make_not_null(&box),
+              [&inbox](const gsl::not_null<db::item_type<tag>*> destination,
+                       TimeStepId time) noexcept {
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/ReceiveWorldtubeData.hpp]: `const TimeStepId&`

File: src/Evolution/Systems/Cce/Actions/ReceiveWorldtubeData.hpp [hunk 14]
```diff
@@ -27,52 +29,97 @@ namespace Actions {
...
+                *destination = get<tag>(*inbox[time].begin());
+              },
+              db::get<::Tags::TimeStepId>(box));
+        });
+    const auto& time_step_id = db::get<::Tags::TimeStepId>(box);
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/ReceiveWorldtubeData.hpp]: combine these two lines

File: src/Evolution/Systems/Cce/ReceiveTags.hpp [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+#pragma once
+
+#include "DataStructures/Variables.hpp"
+#include "Time/TimeStepId.hpp"
+
+namespace Cce {
+namespace ReceiveTags {
+
+namespace detail {
+struct AllDifferent {
```
Comment (kidder) [inline src/Evolution/Systems/Cce/ReceiveTags.hpp]: @nilsdeppe will put in a quick PR (hopefully today) making inserting something into an inbox more flexible so we can get rid of this struct...

**Review by kidder — CHANGES_REQUESTED** [review]
(with @nilsdeppe ) you can squash

File: src/Evolution/Systems/Cce/ReceiveTags.hpp (hunk 15 above)
Comment (moxcodes) [reply src/Evolution/Systems/Cce/ReceiveTags.hpp]: that sounds lovely. I was frustrated by the constraints that led to this workaround.

**Comment by moxcodes** [comment]
Squashed.
Cheers!

File: tests/Unit/Evolution/Systems/Cce/Actions/Test_RequestBoundaryData.cpp [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+  template <typename ParallelComponent, typename... DbTags,
+            typename Metavariables, typename ArrayIndex,
+            Requires<tmpl2::flat_any_v<cpp17::is_same_v<
+                ::Tags::Variables<
+                    typename Metavariables::cce_boundary_communication_tags>,
+                DbTags>...>> = nullptr>
+  static void apply(const db::DataBox<tmpl::list<DbTags...>>& /*box*/,
+                    const Parallel::ConstGlobalCache<Metavariables>& /*cache*/,
+                    const ArrayIndex& /*array_index*/,
+                    const TimeStepId time) noexcept {
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/Cce/Actions/Test_RequestBoundaryData.cpp]: sorry, just noticed this should also be a `const&`

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by moxcodes** [comment]
residual fixup added.
Cheers!

**Review by kidder — COMMENTED** [review]
squash, thanks!

**Comment by moxcodes** [comment]
Squashed.
Cheers!

**Review by kidder — DISMISSED** [review]

File: src/Evolution/Systems/Cce/ReceiveTags.hpp (hunk 15 above)
Comment (nilsdeppe) [reply src/Evolution/Systems/Cce/ReceiveTags.hpp]: I've submitted #1989, which should make this a lot easier (hopefully just having the inbox tag inherit off of one of the inserters added there)

**Comment by nilsdeppe** [comment]
@moxcodes can you ping us once this is rebased with the new inbox tags? :)

**Comment by moxcodes** [comment]
rebased with the new inbox tags. The only commit changed was `Add communication...`, and only with respect to the receive tags and unpacking the inbox.

**Review by nilsdeppe — APPROVED** [review]

**Review by kidder — APPROVED** [review]
